pkg/archive/archive.go

```diff
@@ -349,7 +349,13 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
 
 	// There is no LChmod, so ignore mode for symlink. Also, this
 	// must happen after chown, as that can modify the file mode
-	if hdr.Typeflag != tar.TypeSymlink {
+	if hdr.Typeflag == tar.TypeLink {
+		if fi, err := os.Lstat(hdr.Linkname); err == nil && (fi.Mode()&os.ModeSymlink == 0) {
```

Comment:
Here is checked not the hardlink itself, but the file, which it points to. It this file is a dangling symlink, then chmod will fail, so we have to handle this case. If "Linkname" is another hardlink, it's ok, because on disk it's transparently casted to regular file or symlink.